dissect/target/plugin.py

```diff
@@ -71,7 +63,111 @@ class OperatingSystem(StrEnum):
     WINDOWS = "windows"
 
 
-def export(*args, **kwargs) -> Callable:
+@dataclass(frozen=True, eq=True)
+class PluginDescriptor:
+    module: str
+    qualname: str
+    namespace: str
+    path: str
+    findable: bool
+    functions: list[str]
+    exports: list[str]
+
+
+@dataclass(frozen=True, eq=True)
+class FunctionDescriptor:
+    name: str
+    namespace: str
+    path: str
+    exported: bool
+    internal: bool
+    findable: bool
+    output: str | None
+    method_name: str
+    module: str
+    qualname: str
+
+
+@dataclass(frozen=True, eq=True)
+class FailureDescriptor:
+    module: str
+    stacktrace: list[str]
+
+
+PluginDescriptorLookup = TypedDict(
+    "PluginDescriptorLookup",
+    {
+        None: dict[str, PluginDescriptor],
+        "__os__": dict[str, PluginDescriptor],
+        "__child__": dict[str, PluginDescriptor],
+    },
+)
+
+
+FunctionDescriptorLookup = TypedDict(
+    "FunctionDescriptorLookup",
+    {
+        None: dict[str, dict[str, FunctionDescriptor]],
+        "__os__": dict[str, dict[str, FunctionDescriptor]],
+        "__child__": dict[str, dict[str, FunctionDescriptor]],
+    },
+)
+
+
+PluginRegistry = TypedDict(
+    "PluginRegistry",
+    {
+        "__plugins__": PluginDescriptorLookup,
+        "__functions__": FunctionDescriptorLookup,
+        "__ostree__": dict[str, PluginDescriptor],
+        "__failed__": list[FailureDescriptor],
+    },
+)
+
+PLUGINS: PluginRegistry = {
+    # Plugin descriptor lookup
+    # {"<plugin index>": {"<module_path>": PluginDescriptor}}
+    "__plugins__": {
+        # All regular plugins
+        # {"<module_path>": PluginDescriptor}
+        None: {},
+        # All OS plugins
+        # {"<module_path>": PluginDescriptor}
+        "__os__": {},
+        # All child plugins
+        # {"<module_path>": PluginDescriptor}
+        "__child__": {},
```

Comment:
(Nitpick)

Discriminating between child and regular plugins is done for performance it seems? Perhaps somewhat better to  localize the distinction and add memoization to  `child_plugins`.

Reply:
There's no (fast) way to distinct a `ChildTargetPlugin` from a regular plugin once they're in the registry. We only allow basic types in the registry so that it can be easily serialised. Child plugins are special cased and only used in the opening on child targets, so they can't be used in the same places that regular plugins can.

I don't think there's a much better way to mix these.